python/lsst/daf/butler/registry/collections/_base.py

```diff
@@ -456,8 +455,49 @@ def __getitem__(self, key: Any) -> CollectionRecord:
         except KeyError as err:
             raise MissingCollectionError(f"Collection with key '{key}' not found.") from err
 
-    def __iter__(self) -> Iterator[CollectionRecord]:
-        yield from self._records.values()
+    def resolve_wildcard(
+        self,
+        wildcard: CollectionWildcard,
+        *,
+        collection_types: Set[CollectionType] = CollectionType.all(),
+        done: set[str] | None = None,
+        flatten_chains: bool = True,
+        include_chains: bool | None = None,
+    ) -> list[CollectionRecord]:
+        # Docstring inherited
+        if done is None:
+            done = set()
+        include_chains = include_chains if include_chains is not None else not flatten_chains
+
+        def resolve_nested(record: CollectionRecord, done: set[str]) -> Iterator[CollectionRecord]:
+            if record.name in done:
+                return
+            if record.type in collection_types:
+                done.add(record.name)
+                if record.type is not CollectionType.CHAINED or include_chains:
+                    yield record
+            if flatten_chains and record.type is CollectionType.CHAINED:
+                done.add(record.name)
+                for name in cast(ChainedCollectionRecord, record).children:
+                    # flake8 can't tell that we only delete this closure when
+                    # we're totally done with it.
+                    yield from resolve_nested(self.find(name), done)  # noqa: F821
+
+        result: list[CollectionRecord] = []
+
+        if wildcard.patterns is Ellipsis:
+            for record in self._records.values():
+                result.extend(resolve_nested(record, done))
+            del resolve_nested
+            return result
+        for name in wildcard.strings:
+            result.extend(resolve_nested(self.find(name), done))
+        if wildcard.patterns:
+            for record in self._records.values():
+                if any(p.fullmatch(record.name) for p in wildcard.patterns):
+                    result.extend(resolve_nested(record, done))
+        del resolve_nested
```

Comment:
Is this needed right before return?

Reply:
This is something I've started doing on @natelust 's advice: apparently Python is bad at recognizing when a nested function needs to survive after it goes out of scope - it goes into some sort of delayed garbage-collection channel.  And explicitly deleting the function before return works around that.